compiler/damlc/daml-stdlib-src/DA/Set.daml

```diff
@@ -0,0 +1,146 @@
+-- Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+{-# LANGUAGE CPP #-}
+
+
+#ifndef DAML_GENMAP
+
+-- | HIDE
+module DA.Set where
+
+#else
+
+-- | This module exports the generic set type `Set k` and associated
+-- functions. This module should be imported qualified, for example:
+--
+-- ```
+-- import DA.Set (Set)
+-- import DA.Set qualified as S
+-- ```
+--
+-- This will give access to the `Set` type, and the various operations
+-- as `S.lookup`, `S.insert`, `S.fromList`, etc.
+--
+-- `Set k` internally uses the built-in order for the type `k`.
+-- This means that keys that contain functions are not comparable
+-- and will result in runtime errors. To prevent this, the `Ord k`
+-- instance is required for most set operations. It is recommended to
+-- only use `Set k` for key types that have an `Ord k` instance
+-- that is derived automatically using `deriving`:
+--
+-- ```
+-- data K = ...
+--   deriving (Eq, Ord)
+-- ```
+--
+-- This includes all built-in types that aren't function types, such as
+-- `Int`, `Text`, `Bool`, `(a, b)` assuming `a` and `b` have default
+-- `Ord` instances, `Optional t` and `[t]` assuming `t` has a
+-- default `Ord` instance, `Map k v` assuming `k` and `v` have
+-- default `Ord` instances, and `Set k` assuming `k` has a
+-- default `Ord` instance.
+module DA.Set
+  ( Set
+  , empty
+  , size
+  , toList
+  , fromList
+  , toMap
+  , fromMap
+  , member
+  , null
+  , insert
+  , filter
+  , delete
+  , singleton
+  , union
+  , intersection
+  , difference
+  ) where
+
+import DA.Map (Map)
+import DA.Map qualified as M
+
+-- | The type of a set. This is a wrapper over the `Map` type.
+data Set k = Set with map : Map k ()
```

Comment:
We should move this to a stable type. iirc the only reason why we didn’t do it for `DA.Next.Set` was that it was soon going to be deprecated which turned out to be not quite as soon :grin: Doesn’t have to be in this PR

Reply:
Good point!